src/GraphQL/Output.hs

```diff
@@ -0,0 +1,76 @@
+-- | GraphQL output.
+--
+-- How we encode GraphQL responses.
+module GraphQL.Output
+  ( Response(..)
+  , Errors
+  , Error(..)  -- XXX: Maybe export helper functions rather than constructors.
+  ) where
+
+import Protolude hiding (Location, Map)
+
+import Data.List.NonEmpty (NonEmpty)
+import qualified Data.Map as Map
+import GraphQL.Value (Map, ToValue(..), Value(ValueNull))
+
+-- | GraphQL response.
+--
+-- A GraphQL response must:
+--
+--   * be a map
+--   * have a "data" key iff the operation executed
+--   * have an "errors" key iff the operation encountered errors
+--   * not include "data" if operation failed before execution (e.g. syntax errors,
+--     validation errors, missing info)
+--   * not have keys other than "data", "errors", and "extensions"
+--
+-- Other interesting things:
+--
+--   * Doesn't have to be JSON, but does have to have maps, strings, lists,
+--     and null
+--   * Can also support bool, int, enum, and float
+--   * Value of "extensions" must be a map
+--
+-- "data" must be null if an error was encountered during execution that
+-- prevented a valid response.
+--
+-- "errors"
+--
+--   * must be a non-empty list
+--   * each error is a map with "message", optionally "locations" key
+--     with list of locations
+--   * locations are maps with 1-indexed "line" and "column" keys.
+data Response
+  = Success Map
+  | PreExecutionFailure Errors
+  | ExecutionFailure Errors
+  | PartialSuccess Map Errors
```

Comment:
I'm pretty sure it does make sense (although maybe my comment doesn't). 

> If the data entry in the response is not null, the errors entry in the response may contain any errors that occurred during execution. If errors occurred during execution, it should contain those errors.

From https://facebook.github.io/graphql/#sec-Errors

Note that `data` can be absent or null, and there are different meanings associated with each.